File: perl/examples/list_student_courses.pl
#!/usr/bin/perl -w
# written by andrewt@cse.unsw.edu.au as a 2041 lecture example
# print the courses being taken by each student enrolled in the specified courses

$enrollment_file = shift @ARGV or die;
$debug = 0;

open C, "course_codes" or die "$0: can not open course_codes: $!";
while ($line = <C>) {
    chomp $line;
    $line =~ s/^\s+//;
    $code = substr($line, 0, 8);
    $name = substr($line, 9);
    $course_name{$code} = $name;
    print STDERR "code='$code' -> name='$name'\n" if $debug;
}

open F, "<$enrollment_file" or die;
while (<F>) {
    ($course,$upi,$name) = split /\s*\|\s*/;
    $course{$upi} .=  " $course";
    $name =~ s/(.*), (.*)/$2 $1/;
    $name =~ s/ .* / /;
    $name{$upi} = $name;
}

foreach $course (@ARGV) {
    foreach $upi (keys %course) {
        $courses = $course{$upi};
        next if $courses !~ /$course/;
        print "$name{$upi} is taking:\n";
        $courses =~ s/^ //;
        @courses = split / /, $courses;
        foreach $course (@courses) {
            print "$course $course_name{$course}\n";
        }
    }
}

File: python/examples/list_student_courses.py
#!/usr/bin/python
# written by andrewt@cse.unsw.edu.au as a 2041 lecture example
# print the courses being taken by each student enrolled in the specified courses

import sys,re
if len(sys.argv) < 3:
    print >>sys.stderr, "Usage: %s <enrollment_file> <course_codes>" % sys.argv[0]
    sys.exit(1)
enrollment_file = sys.argv.pop(1)

course_names = {}
for line in open("course_codes"):
    line = line.rstrip()
    code = line[0:8]
    name = line[9:]
    course_names[code] = name

courses = {}
names = {}
for line in open(enrollment_file):
    (course,upi,name) = line.split("|")[0:3]
    m = re.match(r'(.*),\s+(.*\S)', name)
    if m:
    	name = m.group(2) + " " + m.group(1)
    if upi in courses:
        courses[upi] += " " + course
    else:
        courses[upi] = course
    names[upi] = name.rstrip()
    

for course in sys.argv[1:]:
    for upi in courses.keys():
        student_courses = courses[upi].split()
        if course not in student_courses:
            continue
        print "%s is taking"%(names[upi])
        for course in student_courses:
            print "%s %s"%(course, course_names[course])
